// cds-lib/test/cds-lib_test.js
'use strict';

var cdsLib = require('../index.js');

/*
  ======== A Handy Little Nodeunit Reference ========
  https://github.com/caolan/nodeunit

  Test methods:
    test.expect(numAssertions)
    test.done()
  Test assertions:
    test.ok(value, [message])
    test.equal(actual, expected, [message])
    test.notEqual(actual, expected, [message])
    test.deepEqual(actual, expected, [message])
    test.notDeepEqual(actual, expected, [message])
    test.strictEqual(actual, expected, [message])
    test.notStrictEqual(actual, expected, [message])
    test.throws(block, [error], [message])
    test.doesNotThrow(block, [error], [message])
    test.ifError(value)
*/

exports.cdsLib = {
  setUp: function(done) {
    // setup here
    done();
  },
  'update': function(test) {

    test.expect(1);
    // tests here
    var content = {
      attributes: {
        location: 'sales/demo/cds-cli/example1.html',
        id: '5352f008f778fc389837862c'
      },
      opts: {},
      body: 'DEMO TESTING AGAIN'
    };
    cdsLib.upload(content, function(err, uploaded) {
      test.ok(uploaded, 'should be uploaded.');
      test.done();
    });

  },
  'new version with name set': function(test) {

    test.expect(1);
    // tests here
    var content = {
      attributes: {
        location: 'sales/demo/cds-cli/example1.html',
        id: false,
        name: 'CDS Test Version'
      },
      opts: {},
      body: 'DEMO TESTING'
    };
    cdsLib.upload(content, function(err, uploaded) {
      test.ok(uploaded, 'should be uploaded.');
      test.done();
    });

  },
  'new version without name set': function(test) {

    test.expect(1);
    // tests here
    var content = {
      attributes: {
        location: 'sales/demo/cds-cli/example1.html',
        id: false
      },
      opts: {},
      body: 'DEMO TESTING'
    };
    cdsLib.upload(content, function(err, uploaded) {
      test.ok(uploaded, 'should be uploaded.');
      test.done();
    });

  }

};
